drivers/video/video_esp32_dvp.c

```diff
@@ -0,0 +1,440 @@
+
+	const cam_hal_config_t hal_cfg = {
+		.port = 0,
+		.byte_swap_en = cfg->invert_byte_order,
+	};
+
+	cam_hal_init(&data->hal, &hal_cfg);
+
+	cam_ll_reverse_dma_data_bit_order(data->hal.hw, cfg->invert_bit_order);
+	cam_ll_enable_invert_pclk(data->hal.hw, cfg->invert_pclk);
+	cam_ll_set_input_data_width(data->hal.hw, cfg->data_width);
+	cam_ll_enable_invert_de(data->hal.hw, cfg->invert_de);
+	cam_ll_enable_invert_vsync(data->hal.hw, cfg->invert_vsync);
+	cam_ll_enable_invert_hsync(data->hal.hw, cfg->invert_hsync);
+}
+
+static int esp32_init(const struct device *dev)
+{
+	const struct video_esp32_config *cfg = dev->config;
+	struct video_esp32_data *data = dev->data;
+	int ret = 0;
+
+	data->config = cfg;
+
+	k_fifo_init(&data->fifo_in);
+	k_fifo_init(&data->fifo_out);
+
+	video_esp32_cam_ctrl_init(dev);
+
+	if (!device_is_ready(cfg->dma_dev)) {
+		return -ENODEV;
+	}
+
+	LOG_DBG("esp32 video driver loaded");
+
+	return ret;
+}
+
+static const struct video_driver_api esp32_driver_api = {
+	/* mandatory callbacks */
+	.set_format = video_esp32_set_fmt,
+	.get_format = video_esp32_get_fmt,
+	.stream_start = video_esp32_stream_start,
+	.stream_stop = video_esp32_stream_stop,
+	.get_caps = video_esp32_get_caps,
+	/* optional callbacks */
+	.enqueue = video_esp32_enqueue,
+	.dequeue = video_esp32_dequeue,
+	.flush = NULL,
+	.set_ctrl = video_esp32_set_ctrl,
+	.get_ctrl = video_esp32_get_ctrl,
+	.set_signal = NULL,
+};
+
+PINCTRL_DT_INST_DEFINE(0);
+
+static const struct video_esp32_config esp32_config = {
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
+	.source_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0, source)),
+	.dma_dev = ESP32_DT_INST_DMA_CTLR(0, rx),
+	.rx_dma_channel = DT_INST_DMAS_CELL_BY_NAME(0, rx, channel),
+	.data_width = DT_INST_PROP_OR(0, data_width, 8),
+	.invert_bit_order = DT_INST_PROP(0, invert_bit_order),
+	.invert_byte_order = DT_INST_PROP(0, invert_byte_order),
+	.invert_pclk = DT_INST_PROP(0, invert_pclk),
+	.invert_de = DT_INST_PROP(0, invert_de),
+	.invert_hsync = DT_INST_PROP(0, invert_hsync),
+	.invert_vsync = DT_INST_PROP(0, invert_vsync),
+	.cam_clk = DT_INST_PROP_OR(0, cam_clk, 0),
+	.clock_dev = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(0)),
+	.clock_subsys = (clock_control_subsys_t)DT_INST_CLOCKS_CELL(0, offset),
+};
+
+static struct video_esp32_data esp32_data = {0};
+
+DEVICE_DT_INST_DEFINE(0, &esp32_init, PM_DEVICE_DT_INST_GET(idx), &esp32_data, &esp32_config,
+		      POST_KERNEL, CONFIG_VIDEO_INIT_PRIORITY, &esp32_driver_api);
+
+static int video_esp32_cam_init_master_clock(void)
+{
+	int ret = 0;
+
+	ret = pinctrl_apply_state(esp32_config.pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		printk("video pinctrl setup failed (%d)", ret);
+		return ret;
+	}
+
+	/* Enable peripheral */
+	if (!device_is_ready(esp32_config.clock_dev)) {
+		return -ENODEV;
+	}
+
+	clock_control_on(esp32_config.clock_dev, esp32_config.clock_subsys);
+
+	if (!esp32_config.cam_clk) {
+		printk("No cam_clk specified\n");
```

Comment:
I suggest using `LOG_ERR` instead of `printk`.

Reply:
This function runs on `PRE_KERNEL_2`, logging is not yet available :/